engine/env/spark/src/main/java/org/datacleaner/spark/SparkJobContext.java

```diff
@@ -122,16 +126,73 @@ public String getAnalysisJobPath() {
     }
 
     public String getComponentKey(ComponentJob componentJob) {
-        return LabelUtils.getLabel(componentJob);
+        List<ComponentJob> componentJobList = getComponentList();
+        for (int i = 0; i < componentJobList.size(); i++) {
+            if (componentJob.equals(componentJobList.get(i))) {
+                return String.valueOf(i);
+            }
+        }
+        return null;
     }
     
     public ComponentJob getComponentByKey(String key) {
-        final List<AnalyzerJob> analyzerJobs = getAnalysisJob().getAnalyzerJobs();
-        for (AnalyzerJob analyzerJob : analyzerJobs) {
-            if (key.equals(LabelUtils.getLabel(analyzerJob))) {
-                return analyzerJob;
+        List<ComponentJob> componentJobList = getComponentList();
+        for (int i = 0; i < componentJobList.size(); i++) {
+            final ComponentJob componentJob = componentJobList.get(i);
+            if (key.equals(getComponentKey(componentJob))) {
```

Comment:
It seems that getComponentKey(ComponentJob) is also calling getComponentList() which is not a very quick operation to do I can imagine (quite some graph traversal in it). Maybe we should provide a overridden method where you can also pass in the component list so that this can be done just once.